lib/linguist/languages.yml

```diff
@@ -4304,6 +4304,16 @@ Python traceback:
   tm_scope: text.python.traceback
   ace_mode: text
   language_id: 304
+Q#:
+  type: programming
+  extensions:
+  - ".qs"
```

Comment:
Are [these files](https://github.com/search?q=extension%3Aqs+NOT+qsharp&type=Code) Q# too?

Reply:
Thanks for pointing that out – some of them are, however, now I see that the ".qs" file extension is also used by Qt Script, a scripting language for Qt based on ECMAScript. The language is not recognized by Linguist, but adding Q# in the current form would most likely classify thousands of [Qt scripts](https://github.com/search?q=extension%3Aqs+NOT+quantum+NOT+qsharp+NOT+q%23&type=Code) as Q#, as it would be the only language officially using the extension.

One way to deal with this would be to also add Qt Script support to Linguist and then add some heuristics to disambiguate. However, the language came to existence around 2007 and no one cared enough to make it recognizable by Linguist for nearly 13 years – there must be a good reason why.

Is it possible to define heuristics such that those ".qs" files that pass are recognized as Q#, and everything else is classified as no language (just like before)? That'd make things much simpler.

Reply:
> However, the language came to existence around 2007 and no one cared enough to make it recognizable by Linguist for nearly 13 years – there must be a good reason why.

I wouldn't make too much of that. Old languages get added as frequently as recent languages.

> Is it possible to define heuristics such that those ".qs" files that pass are recognized as Q#, and everything else is classified as no language (just like before)?

No. Unfortunately, we don't have a no-language option. That would require some work on our default strategy, the Bayesian classifier, and none of the core contributors have had time to tackle it yet.

> One way to deal with this would be to also add Qt Script support to Linguist and then add some heuristics to disambiguate.

I think that's the best way forward.

Reply:
Should I create a separate PR introducing Qt Script support, or rather include it in this one?

Reply:
It's best to add both languages in the present pull request. That way there's no chances of having only one pull request included in the next release (which would mean lots of misdetections until we make a new release).

Reply:
I've added support for Qt Script and introduced some other changes (see the updated PR description). As for the heuristics, the [contribution guide](https://github.com/github/linguist/blob/master/CONTRIBUTING.md) says to ping @lildude to help with testing the classifier on lots and lots of source files, so I'm doing just that now. The classifier is correct when ran locally on the "dataset" consisting of four sample source files :smile: I haven't delved too deep into the inner workings of Linguist - is the fact that Qt Script uses JS's syntax somehow involved in the classification process?

Reply:
Ooof, those docs are out of date 🤦, we've not done that in a while. You can grab a collection yourself using  [Harvester](https://github.com/Alhadis/Harvester#downloading-files).

Ideally, we'd prefer a heuristic rather than relying solely on the classifier as it really is the last resort and often gets it wrong.

Reply:
@lildude I've used Harvester, however, it only managed to scrape 227 Q# and 49 Qt Script files (it breaks with a `SyntaxError` after some time). The interesting part is that Linguist classifies all the 276 files correctly even without any heuristics. I've corrected for file and directory names so that they are not indicative of the language, but still – it just works! I've added some patterns to `heuristics.yml` though, just to be sure. They are based on some mutually exclusive keywords.

@pchaigno I've updated the checkboxes in the description. I can add the syntax for Q# later, when microsoft/qsharp-compiler#504 gets resolved. So, for now, I guess that's it? Are we ready to merge? 😄

Reply:
> it breaks with a `SyntaxError` after some time

Feel free to open an issue for that. 😉

Reply:
Alright, after taking a better look at things, it appears that I've just constructed the query incorrectly 🤷‍♂️ I'll run the classifier on a few thousand samples and fine-tune the heuristics.

Reply:
I've ran Linguist for 2986 files in total. The sample was a bit skewed, this time I got 2544 Q# vs. 442 Qt Script files from Harvester (`harvest("extension:qs quantum")` and `harvest("extension:qs qt script")`), however, all of them were classified properly with the updated heuristics.